### ISSUE TITLE
Multiplayer - two DOSbox instances on the same PC

### ISSUE THREAD

**Issue by Torinde** [dosbox-staging/dosbox-staging#2461]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Windows 10

### Is your feature request related to a problem? Please describe.

It would be great if a multiple instances can run on the same PC and be used by multiple persons simultaneously while sitting next to each other. 

Multiple DOSbox-X instances running on the same computer (big screen 8K/SuperUltraWide/4K regular screens) - for multiplayer.

Input via (some of):
- virtual keyboard/mouse/joystick
- assigning keyboard to 1st instance, mouse to 2nd, joytsick1 to 3rd, joystick2 to 4th, etc.
- assigning PART of keyboard to first instance, another PART to second instance, etc.
- ManyMouse (mouse1 to 1st instance, mouse2 to 2nd, etc.) and customizing keyboard parts in-game (e.g. one instance gets ASDW, the second instance gets the arrows)
- "ManyKeyboard" assigning keyboard1 to 1st instance, keyboard2 to 2nd, etc. @icculus
- USB pass-through, so that each instance can get its own physical keyboard/mouse

Sound output via different audio devices (wired/Bluetooth headphones, etc.) or just mute all but one instances.

Multi-instance coordinated "fullscreens" (or DOSbox window to get black padding when extended outside the selected aspect ratio, so that by manually adjusting the windows the whole screen can become occupied).

SDL3 feature that may help: https://github.com/libsdl-org/SDL/issues/7498

Same topic at: joncampbell123/dosbox-x/issues/4113

@Grounded0, posting here inspired by your [guide](https://github.com/dosbox-staging/dosbox-staging/wiki/Multiplayer-&-serial-ports)

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
Sure, it will be a nice feature to tackle in 2050 when everything more important is already in tip-top shape 😄 See you then 😏

**Comment by kcgen** [comment]
@Torinde , are you aware of any operating systems that let you partition the input from a single device simultaneously among several applications?

- left-side-of keyboard routed to application X
- right-side-of-keyboard routed to application Y

I suspect this issue needs to be resolved first at the OS level. Once it is, then no further work will be needed at the DOSBox level (each instance will get a "view" into their portion of the input device).

**Comment by johnnovak** [comment]
@Torinde these far-out idea dumps are not really helping... The issue tracker is for a) raising bugs, b) creating tickets for work that is more or less well-defined and achievable in the near future. These huge laundry-list / checklist type of tickets of yours are neither well-defined, nor achievable; they're just huge idea-dumps without much regard for practicality, usefulness, or the level of effort needed to implement them.

I've been silent until now, but I'm seeing a pattern here, and I wish you'd stop raising these types of tickets. Have you thought about how much work it would be to implement these as you "specified" them? How much would they complicate the codebase? How many people would be interested in them? Does the level of effort and the increase in complexity of the codebase "pay off" so to speak? Please think a bit more before raising these and be realistic.

**Comment by Torinde** [comment]
@kcgen, I'm not aware of such OS doing split keyboard per application. That's why I was thinking of DOSbox doing it internally, e.g. masking/reserving/not using/ignoring some of the keys in instance1, and the reverse mask in instance2.

@johnnovak, OK, I'll stop sharing those ideas as unwelcome, understanding the team wants to keep the issue tracker "clean", using it for near term/urgent tasks only. I was wondering whether to file that as "discussion" or "issue - enhancement" - my perception was that an "issue - enhancement" can be a long term/far fetched one and there is nothing wrong with it sitting open for years/forever (like an unread email "I'm not deleting that outright, but don't have time for it either"). Wrong, clear.

Pattern - yes, I posted a few here. But if you want to really see a pattern of those - much more can be found at -X issue tracker, and EVEN MORE in my own file, where those are coming from :)

Usefulness - some multiplayer games have co-op modes, sometimes you want to _teach/show_ others, sometimes you want to share the fun and sit together, sometimes you have a single huge screen and no other suitable PCs around, etc.
Emulating (if you're old school): when you bring a desktop computer _and CRT monitor_ to the house of your friend, to connect via serial cable, sometimes the available desk space is right next to his monitor...

**Comment by dreamer** [comment]
> @kcgen, I'm not aware of such OS doing split keyboard per application. That's why I was thinking of DOSbox doing it internally, e.g. masking/reserving/not using/ignoring some of the keys in instance1, and the reverse mask in instance2.

Linux. It's called "multiseat"; I have never used it myself, but there's plenty of info out there, e.g. [1](https://wiki.ubuntu.com/Multiseat), [2](https://wiki.archlinux.org/title/xorg_multiseat).

**Comment by dreamer** [comment]
BTW, @Torinde - I think posting such out-of-the-left-field ideas is not bad, but you must expect that we will sometimes close it as out of scope, so don't get discouraged.

In this specific case - instead of implementing any feature it would be much easier to describe how to accomplish this task using already existing features (in and out of DOSBox) and write a Wiki article - this is something that you can do as well.

Usually, it's a good idea to attach the feature to a specific game and usecase ("I want to play *this* game, which implements multiplayer *this* way, and we could use *this* improvement to make it easier/better to accomplish.").

**Comment by kcgen** [comment]
> > @kcgen, I'm not aware of such OS doing split keyboard per application. That's why I was thinking of DOSbox doing it internally, e.g. masking/reserving/not using/ignoring some of the keys in instance1, and the reverse mask in instance2.
> 
> Linux. It's called "multiseat"; I have never used it myself, but there's plenty of info out there, e.g. [1](https://wiki.ubuntu.com/Multiseat), [2](https://wiki.archlinux.org/title/xorg_multiseat).

Yup! It's a long running configuration in libraries:
 - One PC, but many kids at desks each with a monitor, keyboard, and mouse.
 - Each student gets their own Xorg session.

But for Torinde's request to split a single physical keyboard into logical partitions across multiple instances, I believe, is impossible. (but I suppose in the end - this is something lower-level than DOSBox itself).

**Comment by johnnovak** [comment]
> Anyway, niche-of-a-niche here

We're in agreement on that one 😄

**Comment by kcgen** [comment]
> Ideally software-wise, but maybe will be possible by using three Display inputs (e.g. session1 uses HDMI, session2 uses USB-C1, session3 uses USB-C2)... although I don't know if my screen can do side-by-side with more than two inputs.

What about slim laptops without all those output video ports? Maybe ChromeCast would be applicable, too.


> putting aside that I use Windows

This your third ticket where you've pressed further into theoretic areas without any intent to actually use such features.  John and I have both explicitly told you this is time wasting.  

Please read this, from our terms of service:

![2023-05-11_09-50](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/6c821b92-c27a-4a37-b2b0-77e016d433e2)

The wording is meant to be first person, like an oath. Repeat after me: "I will make concise an actionable points to respect your time as a volunteer."

If I see more niche-in-niche-in-niche time wasting tickets without any intent or for you to even use them, I will unfortunately have to block you from out GitHub site, but I certainly don't want to beacuse I /do/ appreciate you more constructive input.

**Comment by johnnovak** [comment]
Please take @kcgen 's points seriously @Torinde. We like it when you contribute something actually helpful or useful, but reading and spending time on these far-out mega tickets takes up time and energy. Don't know about you, maybe you have lots of free time, but I don't, and want to spend it on productive things. I don't want to get into a habit of outright ignoring your tickets either...

Please only raise something if it's a bug report, or a legitimately achievable new feature in the near future. Keep it small, simple, achievable, and _laser focused_. That's the keyword @Torinde, _laser focus_. No meandering, no 50 edge cases, no big list of what if scenarios. I want _you_ to spend the time mulling over it, trimming do wen the fat, keeping just the essence, and ultimately discarding it if it's not practical to implement. Thanks.

**Comment by Torinde** [comment]
Wow, that escalated quickly. I honestly don't think above reaction was warranted.

I considered this topic closed:
> OK, I'll stop sharing those ideas as unwelcome

I acknowledged my misunderstanding of what you use "Issues" for:
> Wrong, clear.

Then dreamer and kcgen began discussing Linux multiseat. I only reflected on what they wrote ("I believe, is impossible") and shared a way it may be possible. And finished with:

>Anyway, niche-of-a-niche here
Signaling that I still consider the topic closed. Obviously unsuccessfully.

At the same time revoking edit rights you gave me earlier was just an extra signal that my participation is unwelcome. Which is fine - project is yours, you define the way of working, no need to accommodate random strangers.
